janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java

```diff
@@ -782,91 +860,141 @@ public FilterBuilder getFilter(Condition<?> condition, KeyInformation.StoreRetri
+                    return QueryBuilders.prefixQuery(fieldName, (String) value);
                 } else if (janusgraphPredicate == Text.REGEX) {
-                    return FilterBuilders.regexpFilter(fieldName, (String) value);
+                    return QueryBuilders.regexpQuery(fieldName, (String) value);
                 } else if (janusgraphPredicate == Cmp.EQUAL) {
-                    return FilterBuilders.termFilter(fieldName, (String) value);
+                    return QueryBuilders.termQuery(fieldName, (String) value);
                 } else if (janusgraphPredicate == Cmp.NOT_EQUAL) {
-                    return FilterBuilders.notFilter(FilterBuilders.termFilter(fieldName, (String) value));
+                    return QueryBuilders.boolQuery().mustNot(QueryBuilders.termQuery(fieldName, (String) value));
                 } else
                     throw new IllegalArgumentException("Predicate is not supported for string value: " + janusgraphPredicate);
-            } else if (value instanceof Geoshape) {
-                Preconditions.checkArgument(janusgraphPredicate == Geo.WITHIN, "Relation is not supported for geo value: " + janusgraphPredicate);
+            } else if (value instanceof Geoshape && Mapping.getMapping(informations.get(key)) == Mapping.DEFAULT) {
+                // geopoint
                 Geoshape shape = (Geoshape) value;
+                Preconditions.checkArgument(janusgraphPredicate instanceof Geo && janusgraphPredicate != Geo.CONTAINS, "Relation not supported on geopoint types: " + janusgraphPredicate);
+
+                final QueryBuilder queryBuilder;
                 if (shape.getType() == Geoshape.Type.CIRCLE) {
                     Geoshape.Point center = shape.getPoint();
-                    return FilterBuilders.geoDistanceFilter(key).lat(center.getLatitude()).lon(center.getLongitude()).distance(shape.getRadius(), DistanceUnit.KILOMETERS);
+                    queryBuilder = QueryBuilders.geoDistanceQuery(key).lat(center.getLatitude()).lon(center.getLongitude()).distance(shape.getRadius(), DistanceUnit.KILOMETERS);
                 } else if (shape.getType() == Geoshape.Type.BOX) {
                     Geoshape.Point southwest = shape.getPoint(0);
                     Geoshape.Point northeast = shape.getPoint(1);
-                    return FilterBuilders.geoBoundingBoxFilter(key).bottomRight(southwest.getLatitude(), northeast.getLongitude()).topLeft(northeast.getLatitude(), southwest.getLongitude());
-                } else
+                    queryBuilder = QueryBuilders.geoBoundingBoxQuery(key).bottomRight(southwest.getLatitude(), northeast.getLongitude()).topLeft(northeast.getLatitude(), southwest.getLongitude());
+                } else if (shape.getType() == Geoshape.Type.POLYGON) {
+                    queryBuilder = QueryBuilders.geoPolygonQuery(key);
+                    for (int i=0; i<shape.size(); i++) {
```

Comment:
Done but note while Geoshape has a get(int i) method it is not a Collection so doesn't support stream directly if that's what you were after.